src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs

```diff
@@ -746,11 +758,43 @@ private void FreeLocals(BoundSequence sequence, LocalSymbol doNotRelease)
 
             foreach (var local in sequence.Locals)
             {
-                if ((object)local != doNotRelease)
-                {
-                    FreeLocal(local);
-                }
+                FreeLocal(local);
+            }
+        }
+
+        /// <summary>
+        /// Defines sequence locals and record them so tht they could be retained for the duration of the encompassing expresson
+        /// Use this when taking a reference of the sequence, which can indirectly refer toany of its locals.
+        /// </summary>
+        private void DefineAndRecordLocals(BoundSequence sequence)
+        {
+            if (sequence.Locals.IsEmpty)
+            {
+                return;
+            }
+
+            _builder.OpenLocalScope();
+
+            foreach (var local in sequence.Locals)
+            {
+                var seqLocal = DefineLocal(local, sequence.Syntax);
+                AddExpressionTemp(seqLocal);
+            }
+        }
+
+        /// <summary>
+        /// Closes the visibility/debug scopes for the sequence locals, but keep the local slots from reuse
+        /// for the duration of the encompassing expresson.
+        /// Use this paired with DefineAndRecordLocals when taking a reference of the sequence, which can indirectly refer to any of its locals.
```

Comment:
This is confusing to me. Are we keeping the local slots or the locals themselves? If we let the locals themselves live past their scope, in what sense is this a scope?